### PR TITLE
Shrink rulenode

### LINKED ISSUE #17280: Shrink RuleNode by two words
RuleNode is a frequently-allocated data structure. It lives in components/style/rule_tree/mod.rs

Right now it's 96 bytes on 64-bit platforms. We can shrink this to 80 quite easily.

We should do the following:
(1) Add a size_of test for RuleNode in tests/unit/stylo/size_of.rs.
(2) Change StyleSource to have a "None" variant, and eliminate the Option<> around StyleSource.
(3) Reuse a sibling pointer for free_count, as described in the comment.

CC @nox @jdm @emilio @SimonSapin

### PR DESCRIPTION
**Pull request by br-lewis**
This shrinks rulenode by one word by removing the Option wrapper around the StyleSource and adding None to StyleSource as an additional variant. The issue mentions shrinking by two words but the free_count one was taken care of in #17368.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #17280

Also tested with `./mach test-unit` and `./mach test-stylo` with no errors reported from either.

- [X] There are tests for these changes

r? @bholley

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17404)

### CODE REVIEW

File: components/style/rule_tree/mod.rs [hunk 1]
```diff
@@ -714,7 +724,7 @@ impl HeapSizeOf for StrongRuleNode {
 
 impl StrongRuleNode {
     fn new(n: Box<RuleNode>) -> Self {
-        debug_assert!(n.parent.is_none() == n.source.is_none());
+        debug_assert!(n.parent.is_none() == ! n.source.is_some());
```
Comment (bholley) [inline components/style/rule_tree/mod.rs]: Nit: unnecessary space after the !.

**Comment by bholley** [comment]
Please squash these. r=me with that.

**Comment by zuwow** [comment]
@bholley: commits have been squashed and I removed that space.

**Comment by bholley** [comment]
@bors-servo r+

Thanks!

File: components/style/rule_tree/mod.rs [hunk 2]
```diff
@@ -714,7 +724,7 @@ impl HeapSizeOf for StrongRuleNode {
 
 impl StrongRuleNode {
     fn new(n: Box<RuleNode>) -> Self {
-        debug_assert!(n.parent.is_none() == n.source.is_none());
+        debug_assert!(n.parent.is_none() == !n.source.is_some());
```
Comment (KiChjang) [inline components/style/rule_tree/mod.rs]: I'm sorry, but how is this change better?

**Comment by bholley** [comment]
> -        debug_assert!(n.parent.is_none() == n.source.is_none());
> +        debug_assert!(n.parent.is_none() == !n.source.is_some());
>
> I'm sorry, but how is this change better?

@KiChjang 

Previously |source| was an Option<>, and now it's a custom enum type. I could have insisted on defining an is_none() on that custom type in addition to an is_some() in order to avoid this line in the diff, but that seemed like nit-picking.

**Comment by KiChjang** [comment]
Yeah, I just realized that after I read the source code, but I would have added the is_none function, since it's just !self.is_some(). It's arguably a bit less hard to follow when reading the code, but I don't really mind.

**Comment by zuwow** [comment]
I had considered that as well but decided it didn't make a ton of sense to add in a function that's only used in a debug assert.
